src/unix/lwt_unix_unix.h

```diff
@@ -3009,38 +3014,200 @@ CAMLprim value lwt_unix_bind_job(value fd, value address)
     return lwt_unix_alloc_job(&job->job);
 }
 
-/* getcwd */
+/** getcwd
+
+    Lwt C stubs come in two varieties, depending on whether the underlying C
+    call is blocking or non-blocking. In all cases, the Lwt wrapper around the C
+    call must be made to appear *non*-blocking.
+
+    1. The simple case is when the underlying C call is already non-blocking. An
+       example of this is `lwt_unix_read`, which is used by Lwt to perform reads
+       from file descriptors that are in non-blocking mode. This stub is a
+       simple wrapper around `read(2)`. It converts its arguments from OCaml
+       runtime representation to normal C, machine representation, passes them
+       to `read(2)`, converts the result back to OCaml, and returns.
+
+    2. In case the underlying C call is blocking, as `getcwd(3)` is, Lwt
+       "converts" it to a non-blocking call by running it inside a worker
+       thread. The rest of this comment is concerned with such blocking calls.
+
+    For background on writing C stubs in OCaml, see
+
+      http://caml.inria.fr/pub/docs/manual-ocaml/intfc.html
+
+
+    Each Lwt stub for a blocking C call defines a *job* for the Lwt worker
+    thread pool. The actual thread pool is implemented in `lwt_unix_stubs.c` and
+    is beyond the scope of this comment. It is not necessary to understand it to
+    implement Lwt jobs (indeed, the author currently doesn't remember exactly
+    how it works!).
+
+    A job is defined by providing three functions. Let's say we want to call our
+    new kind of job "`FOO`":
+
+    - `lwt_unix_FOO_job` allocates the job `struct` (see next paragraph).
+    - `worker_FOO` is later called in a worker thread to actually run the job.
+    - `result_FOO` is, even later, called in the main thread to return the
+      result of the job to OCaml, and deallocate the job.
+
+    It is also necessary to define a `struct` with name `struct job_FOO`. This
+    is the representation of the job that will be manipulated by the thread
+    pool. It has several purposes:
+
+    - Store the pointers to `worker_FOO` and `result_FOO`, so the thread pool is
+      able to run them.
+    - Store the C call arguments, or references to them, so they can be accessed
+      by `worker_FOO` when it runs in the worker thread.
+    - Store the C call results, so they can be accessed by `result_FOO` in the
+      main thread.
+    - Be something that can be placed in queues, or manipulated otherwise.
+
+    So, there are three functions and a `struct` to define. It is important that
+    their names fit the pattern given above, because it is expected by Lwt job
+    helper macros.
```

Comment:
I'll give a stab at writing something.

I think that introduces the struct first also makes sense because we are addressing OCaml programmers which are used to type-before-functions.